Exception in transformation with unused variables

While this correctly compiles:
```scala
object Main extends App {
  val MAX = 10
  val s1, s2, target = new Array[Long](MAX)

  var i = 0

  while (i < MAX) {
    target(i) = s1(i) + s2(i)
    i+= 1
  }
}
```

simply adding another unused ```var``` crashes the compiler
```scala
object Main extends App {
  val MAX = 10
  val s1, s2, target = new Array[Long](MAX)

  var i, j = 0

  while (i < MAX) {
    target(i) = s1(i) + s2(i)
    i+= 1
  }
}
```

the output log is this:
```
exception while transforming Main.i = Main.i.+(1) of class class dotty.tools.dotc.ast.Trees$Assign # 8267
exception while transforming {
  Main.target.update(Main.i, Main.s1.apply(Main.i).+(Main.s2.apply(Main.i)))
  Main.i = Main.i.+(1)
} of class class dotty.tools.dotc.ast.Trees$Block # 8268
exception while transforming {
  {
    Main.target.update(Main.i, Main.s1.apply(Main.i).+(Main.s2.apply(Main.i)))
    Main.i = Main.i.+(1)
  }
  while$()
} of class class dotty.tools.dotc.ast.Trees$Block # 8270
exception while transforming if Main.i.<(Main.MAX) then {
  {
    Main.target.update(Main.i, Main.s1.apply(Main.i).+(Main.s2.apply(Main.i)))
    Main.i = Main.i.+(1)
  }
  while$()
} else () of class class dotty.tools.dotc.ast.Trees$If # 8271
exception while transforming def while$(): Unit = 
  if Main.i.<(Main.MAX) then {
    {
      Main.target.update(Main.i, Main.s1.apply(Main.i).+(Main.s2.apply(Main.i)))
      Main.i = Main.i.+(1)
    }
    while$()
  } else () of class class dotty.tools.dotc.ast.Trees$DefDef # 8272
exception while transforming {
  def while$(): Unit = 
    if Main.i.<(Main.MAX) then {
      {
        Main.target.update(Main.i, 
          Main.s1.apply(Main.i).+(Main.s2.apply(Main.i))
        )
        Main.i = Main.i.+(1)
      }
      while$()
    } else ()
  while$()
} of class class dotty.tools.dotc.ast.Trees$Block # 8274
exception while transforming () extends Object() with App { this: Main.type => 
  val MAX: Int = 10
  val s1: Array[Long] = dotty.runtime.Arrays.newLongArray(Main.MAX)
  val s2: Array[Long] = dotty.runtime.Arrays.newLongArray(Main.MAX)
  val target: Array[Long] = dotty.runtime.Arrays.newLongArray(Main.MAX)
  var i: Int = 0
  var j: Int = 0
  {
    def while$(): Unit = 
      if Main.i.<(Main.MAX) then {
        {
          Main.target.update(Main.i, 
            Main.s1.apply(Main.i).+(Main.s2.apply(Main.i))
          )
          Main.i = Main.i.+(1)
        }
        while$()
      } else ()
    while$()
  }
} of class class dotty.tools.dotc.ast.Trees$Template # 8275
exception while transforming final module class Main$() extends Object() with App { this: Main.type => 
  val MAX: Int = 10
  val s1: Array[Long] = dotty.runtime.Arrays.newLongArray(Main.MAX)
  val s2: Array[Long] = dotty.runtime.Arrays.newLongArray(Main.MAX)
  val target: Array[Long] = dotty.runtime.Arrays.newLongArray(Main.MAX)
  var i: Int = 0
  var j: Int = 0
  {
    def while$(): Unit = 
      if Main.i.<(Main.MAX) then {
        {
          Main.target.update(Main.i, 
            Main.s1.apply(Main.i).+(Main.s2.apply(Main.i))
          )
          Main.i = Main.i.+(1)
        }
        while$()
      } else ()
    while$()
  }
} of class class dotty.tools.dotc.ast.Trees$TypeDef # 8276
exception while transforming package <empty> {
  final lazy module val Main: Main$ = new Main$()
  final module class Main$() extends Object() with App { this: Main.type => 
    val MAX: Int = 10
    val s1: Array[Long] = dotty.runtime.Arrays.newLongArray(Main.MAX)
    val s2: Array[Long] = dotty.runtime.Arrays.newLongArray(Main.MAX)
    val target: Array[Long] = dotty.runtime.Arrays.newLongArray(Main.MAX)
    var i: Int = 0
    var j: Int = 0
    {
      def while$(): Unit = 
        if Main.i.<(Main.MAX) then {
          {
            Main.target.update(Main.i, 
              Main.s1.apply(Main.i).+(Main.s2.apply(Main.i))
            )
            Main.i = Main.i.+(1)
          }
          while$()
        } else ()
      while$()
    }
  }
} of class class dotty.tools.dotc.ast.Trees$PackageDef # 8277
```

Hope this tracking is useful!

@andreaTP thx for reporting the bug.

Compiler dies during the Getters phase. Actual stacktrace: 

```
exception occurred while compiling i1216.scala
Exception in thread "main" java.lang.AssertionError: assertion failed: getter i in object Main$
	at scala.Predef$.assert(Predef.scala:165)
	at dotty.tools.dotc.ast.tpd$TreeOps$.becomes$extension(tpd.scala:773)
	at dotty.tools.dotc.transform.Getters.transformAssign(Getters.scala:74)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.goAssign(TreeTransform.scala:669)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transformUnnamed(TreeTransform.scala:1080)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1214)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1206)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:131)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:54)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transform(TreeTransform.scala:1205)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transformUnnamed(TreeTransform.scala:1087)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1214)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1206)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:131)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:54)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transform(TreeTransform.scala:1205)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.dotty$tools$dotc$transform$TreeTransforms$TreeTransformer$$transformStat$1(TreeTransform.scala:1241)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$38.apply(TreeTransform.scala:1243)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$38.apply(TreeTransform.scala:1243)
	at dotty.tools.dotc.core.Decorators$ListDecorator$.loop$1(Decorators.scala:50)
	at dotty.tools.dotc.core.Decorators$ListDecorator$.mapconserve$extension(Decorators.scala:66)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transformStats(TreeTransform.scala:1243)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transformUnnamed(TreeTransform.scala:1086)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1214)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1206)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:131)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:54)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transform(TreeTransform.scala:1205)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transformUnnamed(TreeTransform.scala:1095)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1214)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1206)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:131)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:54)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transform(TreeTransform.scala:1205)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transformNamed(TreeTransform.scala:1004)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1213)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1206)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:131)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:54)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transform(TreeTransform.scala:1205)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.dotty$tools$dotc$transform$TreeTransforms$TreeTransformer$$transformStat$1(TreeTransform.scala:1239)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$38.apply(TreeTransform.scala:1243)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$38.apply(TreeTransform.scala:1243)
	at dotty.tools.dotc.core.Decorators$ListDecorator$.loop$1(Decorators.scala:50)
	at dotty.tools.dotc.core.Decorators$ListDecorator$.mapconserve$extension(Decorators.scala:66)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transformStats(TreeTransform.scala:1243)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transformUnnamed(TreeTransform.scala:1086)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1214)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1206)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:131)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:54)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transform(TreeTransform.scala:1205)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.dotty$tools$dotc$transform$TreeTransforms$TreeTransformer$$transformStat$1(TreeTransform.scala:1241)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$38.apply(TreeTransform.scala:1243)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$38.apply(TreeTransform.scala:1243)
	at dotty.tools.dotc.core.Decorators$ListDecorator$.loop$1(Decorators.scala:50)
	at dotty.tools.dotc.core.Decorators$ListDecorator$.mapconserve$extension(Decorators.scala:66)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transformStats(TreeTransform.scala:1243)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transformUnnamed(TreeTransform.scala:1184)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1214)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1206)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:131)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:54)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transform(TreeTransform.scala:1205)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transformNamed(TreeTransform.scala:1011)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1213)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1206)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:131)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:54)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transform(TreeTransform.scala:1205)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.dotty$tools$dotc$transform$TreeTransforms$TreeTransformer$$transformStat$1(TreeTransform.scala:1239)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$38.apply(TreeTransform.scala:1243)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$38.apply(TreeTransform.scala:1243)
	at dotty.tools.dotc.core.Decorators$ListDecorator$.loop$1(Decorators.scala:50)
	at dotty.tools.dotc.core.Decorators$ListDecorator$.mapconserve$extension(Decorators.scala:66)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transformStats(TreeTransform.scala:1243)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transformUnnamed(TreeTransform.scala:1193)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1214)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1206)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:131)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:54)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transform(TreeTransform.scala:1205)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.macroTransform(TreeTransform.scala:563)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.run(TreeTransform.scala:480)
	at dotty.tools.dotc.core.Phases$Phase$$anonfun$runOn$1.apply(Phases.scala:276)
	at dotty.tools.dotc.core.Phases$Phase$$anonfun$runOn$1.apply(Phases.scala:274)
	at scala.collection.immutable.List.map(List.scala:273)
	at dotty.tools.dotc.core.Phases$Phase$class.runOn(Phases.scala:274)
	at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.runOn(TreeTransform.scala:474)
	at dotty.tools.dotc.Run$$anonfun$compileUnits$1$$anonfun$apply$mcV$sp$1.apply(Run.scala:62)
	at dotty.tools.dotc.Run$$anonfun$compileUnits$1$$anonfun$apply$mcV$sp$1.apply(Run.scala:59)
	at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
	at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply$mcV$sp(Run.scala:59)
	at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply(Run.scala:54)
	at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply(Run.scala:54)
	at dotty.tools.dotc.util.Stats$.monitorHeartBeat(Stats.scala:69)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:54)
	at dotty.tools.dotc.Run.compileSources(Run.scala:51)
	at dotty.tools.dotc.Run.compile(Run.scala:35)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:26)
	at dotty.tools.dotc.Driver.process(Driver.scala:124)
	at dotty.tools.dotc.Driver.process(Driver.scala:93)
	at dotty.tools.dotc.Driver.process(Driver.scala:105)
	at dotty.tools.dotc.Driver.main(Driver.scala:132)
	at dotty.tools.dotc.Main.main(Main.scala)
```

--- a/src/dotty/tools/dotc/ast/Desugar.scala
+++ b/src/dotty/tools/dotc/ast/Desugar.scala
@@ -634,8 +634,11 @@ object desugar {
   def makeAnnotated(cls: Symbol, tree: Tree)(implicit ctx: Context) =
     Annotated(untpd.New(untpd.TypeTree(cls.typeRef), Nil), tree)
 
-  private def derivedValDef(named: NameTree, tpt: Tree, rhs: Tree, mods: Modifiers) =
-    ValDef(named.name.asTermName, tpt, rhs).withMods(mods).withPos(named.pos)
+  private def derivedValDef(named: NameTree, tpt: Tree, rhs: Tree, mods: Modifiers)(implicit ctx: Context) = {
+    val vdef = ValDef(named.name.asTermName, tpt, rhs).withMods(mods).withPos(named.pos)
+    val mayNeedSetter = valDef(vdef)
+    mayNeedSetter
+   }
 
   private def derivedDefDef(named: NameTree, tpt: Tree, rhs: Tree, mods: Modifiers) =
     DefDef(named.name.asTermName, Nil, Nil, tpt, rhs).withMods(mods).withPos(named.pos)

--- a/tests/pos/i1216.scala
+++ b/tests/pos/i1216.scala
@@ -1,0 +1,11 @@
+object Main {
+  val MAX = 10
+  val s1, s2, target = new Array[Long](MAX)
+
+  var i, j = 0
+
+  while (i < MAX) {
+    target(i) = s1(i) + s2(i)
+    i+= 1
+  }
+}

--- a/tests/pos/i1216a.scala
+++ b/tests/pos/i1216a.scala
@@ -1,0 +1,11 @@
+object Main {
+  val MAX = 10
+  val s1, s2, target = new Array[Long](MAX)
+
+  var i = 0
+
+  while (i < MAX) {
+    target(i) = s1(i) + s2(i)
+    i+= 1
+  }
+}

